sr_edc_ethercat_drivers/src/sr10.cpp

```diff
@@ -0,0 +1,575 @@
+  if (flashing & !can_packet_acked)
+  {
+    if (can_data_is_ack(can_data))
+    {
+      can_packet_acked = true;
+    }
+  }
+
+  return true;
+}
+
+/**
+  * This function will call the reinitialization function for the boards attached to the CAN bus
+  */
+void SR10::reinitialize_boards()
+{
+  // Reinitialize motors information
+  sr_hand_lib->reinitialize_motors();
+}
+
+/**
+  * Given the identifier for a certain board (motor board/ muscle driver) determines the right value
+  * for the CAN bus and the ID of the board in that CAN bus.
+  *
+  * @param board_id the unique identifier for the board
+  * @param can_bus pointer to the can bus number we want to determine
+  * @param board_can_id pointer to the board id we want to determine
+*/
+void SR10::get_board_id_and_can_bus(int board_id, int *can_bus, unsigned int *board_can_id)
+{
+  // We're using 2 can busses,
+  // if motor id is between 0 and 9, then we're using the can_bus 1
+  // else, we're using the can bus 2.
+  int motor_id_tmp = board_id;
+  if (motor_id_tmp > 9)
+  {
+    motor_id_tmp -= 10;
+    *can_bus = 2;
+  }
+  else
+  {
+    *can_bus = 1;
+  }
+
+  *board_can_id = motor_id_tmp;
+}
+
+/**
+  * New scales to be set on IMU's gyroscope or accelerometer.
+  * Sets flag (imu_scale_gyr_) indicating that a new "scalling" command should be sent to the IMU.
+  *
+  * @param request service request containing the new scaling factor to be set for the IMU's gyroscope or accelerometer
+  * @param response service response (empty in this case)
+  * @param which pointer to a string that selects IMU's gyroscope or accelerometer
+  * @return true if the new scaling factor was set successfully, false otherwise  
+  */
+bool SR10::imu_scale_callback_(sr_robot_msgs::SetImuScale::Request &request,
+                               sr_robot_msgs::SetImuScale::Response &response,
+                               const char *which)
+{
+  if (request.scale == 0 || request.scale == 1 || request.scale == 2)
+  {
+    if (which == "acc")
+    {
+      imu_scale_acc_ = request.scale;
+    }
+    else if (which == "gyr")
+    {
+      imu_scale_gyr_ = request.scale;
+    }
+
+    imu_scale_change_ = true;
+    return true;
+  }
+  else
+  {
+    ROS_WARN_STREAM("Tried to set illegal value: " << (int) request.scale);
+    return false;
+  }
+}
+
+/** 
+  * This funcion reads the ethercat status and fills the imu_state with the relevant values.
+  *
+  * @param status_data pointer to the received EtherCAT Status data
+  */
+void SR10::readImu(ETHERCAT_DATA_STRUCTURE_0250_PALM_EDC_STATUS * status_data)
+{
+  imu_state_->data_.orientation[0] = 0.0; imu_state_->data_.orientation[1] = 0.0;
+  imu_state_->data_.orientation[2] = 0.0; imu_state_->data_.orientation[3] = 1.0;
+
+  double acc_multiplier = 1 << imu_scale_acc_;
+  double gyr_multiplier = 1 << imu_scale_gyr_;
+
+  int zero_catch = 0;
+
+  for (size_t index = 0; index < 2; ++index)
+  {
+    if (status_data->sensors[ACCX + 0] == 0)
```

Comment:
@dg-shadow can you help us make sense of this when you're back? https://github.com/shadow-robot/sr-ros-interface-ethercat/pull/175